### PR TITLE
swamp: Sync sub test with state sync

### LINKED ISSUE #351: service/header: Sync test needs to know at which specific height should the node be synced
Currently, in our sync test nodes are just waiting to be fully synced, without explicitly checking the height of which we consider this to be correct. This is causing the test to be flaky in CI 
Example of failed : [Run #2](https://github.com/celestiaorg/celestia-node/runs/4809450105?check_suite_focus=true)
Example of passed: [Run #3](https://github.com/celestiaorg/celestia-node/runs/4809482324?check_suite_focus=true)

In both scenarios we are trying to catch up with the very fast block production from the core.Client which is not good. 
To resolve this, we need 2 points to be done:
1. Make block production slower (e.g. locally we can produce around 30-40 blocks and in CI this can be close to 200)
2. Have an API(sub or smt like requestByHeight) from node which can notify the synced height

### LINKED ISSUE #370: swamp/tests/sync: Light client continues sync after abrupt stop and then start
Steps:

1. Bridge node is running
2. Wait until block height reached 30
3. Start the light node (sync is starting too) 
4. Stop the light node 
5. Wait until block height reaches 200
6. Start the light node 

Expectation:
Light node should successfully sync till 200 height

### PR DESCRIPTION
**Pull request by Bidon15**
Depends on #397, #428
Resolves #351 #370

### CODE REVIEW

**Comment by Wondertan** [comment]
@Bidon15, You can now rebase it on main and undraft it

**Review by liamsi — COMMENTED** [review]
> Depends on https://github.com/celestiaorg/celestia-node/pull/397

I don't really see the relation to https://github.com/celestiaorg/celestia-node/pull/397
(is one of the SyncState methods used internally by `GetByHeight`? Doesn't look like it)

File: node/tests/sync_test.go [hunk 1]
```diff
@@ -37,17 +37,23 @@ func TestSyncLightWithBridge(t *testing.T) {
 
 	err := bridge.Start(ctx)
 	require.NoError(t, err)
-	time.Sleep(1 * time.Second)
-	assert.False(t, bridge.HeaderServ.IsSyncing())
+	h, err := bridge.HeaderServ.GetByHeight(ctx, 20)
+	require.NoError(t, err)
+	fmt.Println(h.Commit.Hash().String())
```
[the remark this answers is not in the document]
Comment (Bidon15) [reply node/tests/sync_test.go:42]: Oops. Removing

**Comment by Bidon15** [comment]
> > Depends on #397
> 
> I don't really see the relation to #397 (is one of the SyncState methods used internally by `GetByHeight`? Doesn't look like it)

Yes. You are totally right about the `GetByHeight`, which is related to 397, but was migrated to a standalone #428 
I'll adapt the description as well as add the test that is covering #370 that I've removed initially

**Review by liamsi — APPROVED** [review]
👍🏼

File: node/tests/sync_test.go [hunk 2]
```diff
@@ -37,17 +36,79 @@ func TestSyncLightWithBridge(t *testing.T) {
...
+	state := bridge.CoreClient.IsRunning()
+	require.True(t, state)
+	err := bridge.Start(ctx)
+	require.NoError(t, err)
+
+	h, err := bridge.HeaderServ.GetByHeight(ctx, 20)
+	require.NoError(t, err)
+
+	var ch int64 = 20
+	require.EqualValues(t, h.Commit.BlockID.Hash, sw.GetCoreBlockByHeight(ctx, &ch).BlockID.Hash)
```
Comment (Wondertan) [inline node/tests/sync_test.go:90]: Pls, hide this int ugliness inside the `GetCoreBlockByHeight` method. You can pass here an int and make a pointer to it inside the method. Pls fix this everywhere.

File: node/tests/sync_test.go [hunk 3]
```diff
@@ -37,17 +36,79 @@ func TestSyncLightWithBridge(t *testing.T) {
...
+
+	addrs, err := peer.AddrInfoToP2pAddrs(host.InfoFromHost(sw.Network.Host(bridge.Host.ID())))
```
Comment (Wondertan) [inline node/tests/sync_test.go:92]: You can simply pass `bridge.Host` to `InfoFromHost` without `sw.Network.Host`. Pls remove it everywhere.

File: node/tests/sync_test.go [hunk 4]
```diff
@@ -37,17 +36,79 @@ func TestSyncLightWithBridge(t *testing.T) {
 
 	err := bridge.Start(ctx)
 	require.NoError(t, err)
-	time.Sleep(1 * time.Second)
-	assert.False(t, bridge.HeaderServ.IsSyncing())
+	h, err := bridge.HeaderServ.GetByHeight(ctx, 20)
+	require.NoError(t, err)
```
Comment (Wondertan) [inline node/tests/sync_test.go:40]: Pls make the context here is wrapped with a timeout. This is important to follow in every test

File: node/tests/swamp/swamp.go [hunk 5]
```diff
@@ -226,3 +235,37 @@ func (s *Swamp) NewLightNodeWithStore(store node.Store, options ...node.Option)
...
+func (s *Swamp) remove(rn *node.Node, sn []*node.Node) error {
+	if len(sn) == 1 {
+		sn = nil
+		return nil
+	}
+
+	initSize := len(sn)
+	for i := 0; i < len(sn); i++ {
+		if sn[i] == rn {
+			sn = append(sn[:i], sn[i+1:]...)
```
Comment (Wondertan) [inline node/tests/swamp/swamp.go:262]: Here we have a bug. Append in this case will create a new slice with a newly allocated memory behind it and you only set it to `sn` in the local stack, instead, you should set the result here directly into Swamp.LightNode or Swamp.BridgeNodes. 

Ping @vgonkivs for a more in-depth explanation.

**Review by Wondertan — CHANGES_REQUESTED** [review]

File: node/tests/swamp/swamp.go (hunk 5 above)
Comment (Wondertan) [reply node/tests/swamp/swamp.go:262]: @Bidon15 https://github.com/golang/go/wiki/SliceTricks

File: node/tests/sync_test.go (hunk 3 above)
Comment (Bidon15) [reply node/tests/sync_test.go]: Done

File: node/tests/sync_test.go (hunk 2 above)
Comment (Bidon15) [reply node/tests/sync_test.go]: Agreed. Changed the func naming to Hash, too

File: node/tests/sync_test.go (hunk 4 above)
Comment (Bidon15) [reply node/tests/sync_test.go]: I've made the ctxTimeout for GetByHeight only.
wdyt?

File: node/tests/swamp/swamp.go (hunk 5 above)
Comment (Bidon15) [reply node/tests/swamp/swamp.go:264]: done according to ideas above. Thanks!

File: service/header/store_heightsub.go [hunk 6]
```diff
@@ -40,7 +40,7 @@ func (hs *heightSub) SetHeight(height uint64) {
 // It can return errElapsedHeight, which means a requested header was already provided
 // and caller should get it elsewhere.
 func (hs *heightSub) Sub(ctx context.Context, height uint64) (*ExtendedHeader, error) {
-	if hs.Height() >= height || hs.Height() == 0 { // if height is zero - intentionally avoid subscribing
```
Comment (Bidon15) [inline service/header/store_heightsub.go:43]: I don;t know how but this if statement is not letting the light node sub to headers in the specific test 

cc: @Wondertan

File: service/header/service.go [hunk 7]
```diff
@@ -45,7 +45,11 @@ func (s *Service) Stop(context.Context) error {
 	return nil
 }
 
-// TODO(@Bidon15): Issue #351
+// GetByHeight returns the ExtendedHeader by a provided height
+// This is a blocking call until the height is reached or context deadline is exceeded
```
Comment (renaynay) [inline service/header/service.go:49]:
```suggestion
// GetByHeight returns the ExtendedHeader at the given height, blocking
// until header has been processed by the store or context deadline is exceeded.
```

File: service/header/service.go [hunk 8]
```diff
@@ -45,7 +45,11 @@ func (s *Service) Stop(context.Context) error {
...
+func (s *Service) GetByHeight(ctx context.Context, height uint64) (*ExtendedHeader, error) {
```
Comment (renaynay) [inline service/header/service.go:50]: How would this fit into your idea for RPC on header service @Wondertan ?

**Review by renaynay — CHANGES_REQUESTED** [review]

File: node/tests/swamp/swamp.go [hunk 9]
```diff
@@ -226,3 +235,39 @@ func (s *Swamp) NewLightNodeWithStore(store node.Store, options ...node.Option)
 
 	return node
 }
+
+// RemoveNode removes a node from the swamp's node slice
+// this allows reusage of the same var in the test scenario
+// if the user needs to stop and start the same node
+func (s *Swamp) RemoveNode(n *node.Node, t node.Type) error {
```
[the remark this answers is not in the document]
Comment (Bidon15) [reply node/tests/swamp/swamp.go:242]: No. It is shut down before being removed from the swamp slices in the test scenario

File: service/header/service.go (hunk 8 above)
Comment (Wondertan) [reply service/header/service.go:50]: This fits well, we made this together in some pair coding session, AFAIR.

File: node/tests/sync_test.go (hunk 4 above)
Comment (Wondertan) [reply node/tests/sync_test.go]: I am ok with this as long as there is a global context with a timeout. The global context is needed so we don't ever deadlock while running tests locally or on CI.

Comment (Wondertan) [reply node/tests/sync_test.go]: The blocking may happen literally everywhere where context is passed and not only for GetByHeight, so the timeout should be on the root context. Further, it won't really make sense to have a special context timeout for GetByHeight as there will be one in the root.

**Review by Wondertan — APPROVED** [review]
Approving, considering that the last context timeout issue will be fixed.

File: node/tests/sync_test.go (hunk 4 above)
Comment (Bidon15) [reply node/tests/sync_test.go:41]: Ok, I've changed to global ctx with timeout, so resolving and merging 
thanks for input 🙏

**Review by renaynay — APPROVED** [review]

**Pull request merged**